### PR TITLE
Add ValueCurves from PowerSystems to infrasys

### PR DESCRIPTION
**Pull request by jerrypotts**
- Added classes for InputOutputCurve, IncrementalCurve, and AverageRateCurve that call function_data classes to build generic cost curves. These classes are analogous to the structs found in [PowerSystems](https://github.com/NREL-Sienna/PowerSystems.jl/blob/9e6f8a7e073ef7b606d3bc3a32640fcb69789a99/src/models/cost_functions/ValueCurves.jl)
- Added functions to convert between the different value_curves classes (ex: InputOutputToIncremental creates an IncrementalCurve from an existing InputOutputCurve)
- Added functions to function_data to perform calculations for value_curve conversion
- Added additional serialization tests to test_function_data

### CODE REVIEW

**Review by daniel-thom — COMMENTED** [review]
I made a cursory pass and have two minor comments.

File: src/infrasys/exceptions.py [hunk 1]
```diff
@@ -31,3 +31,7 @@ class ISNotStored(ISBaseException):
 
 class ISOperationNotAllowed(ISBaseException):
     """Raised if the requested operation is not allowed."""
+
+
+class ISMethodError(ISBaseException):
```
[the remark this answers is not in the document]
Comment (pesap) [reply src/infrasys/exceptions.py:36]: Yeah, we can use that instead.

File: src/infrasys/function_data.py [hunk 2]
```diff
@@ -65,28 +69,27 @@ def validate_piecewise_linear_x(points: List[XYCoords]) -> List[XYCoords]:
...
 
     if len(x_coords) < 2:
-        raise ValueError("Must specify at least two x-coordinates")
+        raise ValueError("Must specify at least two x-coordinates.")
     if not (
         x_coords == sorted(x_coords)
         or (np.isnan(x_coords[0]) and x_coords[1:] == sorted(x_coords[1:]))
     ):
-        raise ValueError(f"Piecewise x-coordinates must be ascending, got {x_coords}")
+        raise ValueError(f"Piecewise x-coordinates must be ascending, got {x_coords}.")
```
[the remark this answers is not in the document]
Comment (pesap) [reply src/infrasys/function_data.py]: We missed this on the `pyproject.toml`. The updated version include it and make it compliant.

File: src/infrasys/production_variable_cost_curve.py
Comment (pesap) [inline src/infrasys/production_variable_cost_curve.py:1]: Can we call this file`cost_curves.py` instead?

File: src/infrasys/production_variable_cost_curve.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+from infrasys.function_data import LinearFunctionData
+
+
+class ProductionVariableCostCurve(Component):
+    name: Annotated[str, Field(frozen=True)] = ""
+
+
+class CostCurve(ProductionVariableCostCurve):
+    """Direct representation of the variable operation cost of a power plant in currency. Composed
+    of a Value Curve that may represent input-output, incremental, or average rate
```
Comment (pesap) [inline src/infrasys/production_variable_cost_curve.py:14]: Can we follow numpy style here? First sentence the summary of the class.

File: src/infrasys/production_variable_cost_curve.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        function_data=LinearFunctionData(proportional_term=0.0, constant_term=0.0)
+    )
+
+
+class FuelCurve(ProductionVariableCostCurve):
+    """Representation of the variable operation cost of a power plant in terms of fuel (MBTU,
+    liters, m^3, etc.), coupled with a conversion factor between fuel and currency. Composed of
+    a Value Curve that may represent input-output, incremental, or average rate data.
+    The default units for the x-axis are MW and can be specified with `power_units`.
+    """
```
Comment (pesap) [inline src/infrasys/production_variable_cost_curve.py:38]: Same numpy style

File: src/infrasys/value_curves.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        ----------
+        data : IncrementalCurve
+            Original IncrementalCurve for conversion.
+
+        Returns
+        ----------
+        InputOutputCurve
+            InputOutputCurve using either QuadraticFunctionData or PiecewiseStepData.
+        """
+        if isinstance(self.function_data, LinearFunctionData):
```
Comment (pesap) [inline src/infrasys/value_curves.py:90]: Can you use match/case statement instead?

File: tests/test_production_variable_cost_curve.py
Comment (pesap) [inline tests/test_production_variable_cost_curve.py:1]: Rename according to new file name `test_cost_curves`

**Review by pesap — CHANGES_REQUESTED** [review]
LGTM. If you address the changes we can merge it right away.

**Pull request merged**